app/src/main/java/org/thoughtcrime/securesms/components/WaveFormSeekBarView.java

```diff
@@ -0,0 +1,164 @@
+
+  private static final int ANIM_DURATION             = 450;
+  private static final int ANIM_BAR_OFF_SET_DURATION =  12;
+
+  private final Interpolator overshoot        = new OvershootInterpolator();
+  private final Paint        paint            = new Paint();
+  private       float[]      data             = new float[0];
+  private       long         dataSetTime;
+  private       Drawable     progressDrawable;
+  private       boolean      waveMode;
+  
+  @ColorInt private int playedBarColor   = 0xffffffff;
+  @ColorInt private int unplayedBarColor = 0x7fffffff;
+  @Px       private int barWidth;
+
+  public WaveFormSeekBarView(Context context) {
+    super(context);
+    init();
+  }
+
+  public WaveFormSeekBarView(Context context, @Nullable AttributeSet attrs) {
+    super(context, attrs);
+    init();
+  }
+
+  public WaveFormSeekBarView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
+    super(context, attrs, defStyleAttr);
+    init();
+  }
+
+  private void init() {
+    setWillNotDraw(false);
+    
+    paint.setStrokeCap(Paint.Cap.ROUND);
+    paint.setAntiAlias(true);
+
+    progressDrawable = super.getProgressDrawable();
+
+    if (isInEditMode()) {
+      setWaveData(sinusoidalExampleData());
+      dataSetTime = 0;
+    }
+
+    barWidth = getResources().getDimensionPixelSize(R.dimen.wave_form_bar_width);
+  }
+
+  public void setColors(@ColorInt int playedBarColor, @ColorInt int unplayedBarColor, @ColorInt int thumbTint) {
+    this.playedBarColor   = playedBarColor;
+    this.unplayedBarColor = unplayedBarColor;
+
+    getThumb().setColorFilter(thumbTint, PorterDuff.Mode.SRC_IN);
+
+    invalidate();
+  }
+
+  @Override
+  public void setProgressDrawable(Drawable progressDrawable) {
+    this.progressDrawable = progressDrawable;
+    if (!waveMode) {
+      super.setProgressDrawable(progressDrawable);
+    }
+  }
+
+  @Override
+  public Drawable getProgressDrawable() {
+    return progressDrawable;
+  }
+
+  public void setWaveData(@NonNull float[] data) {
+    if (!Arrays.equals(data, this.data)) {
+      this.data        = data;
+      this.dataSetTime = System.currentTimeMillis();
+    }
+    setWaveMode(data.length > 0);
+  }
+
+  public void setWaveMode(boolean waveMode) {
+    this.waveMode = waveMode;
+    super.setProgressDrawable(this.waveMode ? null : progressDrawable);
+    invalidate();
+  }
+
+  @Override
+  protected void onDraw(Canvas canvas) {
+    if (waveMode) {
+      drawWave(canvas);
+    }
+    super.onDraw(canvas);
+  }
+
+  private void drawWave(Canvas canvas) {
+    paint.setStrokeWidth(barWidth);
+
+    int   usableHeight = getHeight() - getPaddingTop() - getPaddingBottom();
+    int   usableWidth  = getWidth() - getPaddingLeft() - getPaddingRight();
+    float midpoint     = usableHeight / 2f;
+    float maxHeight    = usableHeight / 2f - barWidth;
+    float barGap       = (usableWidth - data.length * barWidth) / (float) (data.length - 1);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).